### PR TITLE
Implement retrieving DBMS constraints

### PR DESCRIPTION
**Pull request by sergeymakinen**
| Q             | A
| ------------- | ---
| Is bugfix?    | no
| New feature?  | yes
| Breaks BC?    | no
| Tests pass?   | yes (thanks to [this](https://github.com/sergeymakinen/docker-yii2-php) and [this](https://github.com/sergeymakinen/docker-test-db))
| Fixed issues  | partially #8163, also see #8152, #9003, #13647

This small PR adds:

- [x] DBMS constraints retrieving methods.
- [x] DBMS constraints low-level manipulation methods.
- [x] DMBS constraints `QueryBuilder` tests.
- [x] DMBS constraints `Command` tests.

Please note: 

- This PR is **not** a complete solution for a constraints management. It's a first part that adds all that low-level stuff. I will implement other parts during a 2.0.13 development timeframe.
- Some existing code is outdated or not needed anymore (or both O:). But it requires a discussion and deserves a separate issue.
- The SQL tokenizer is aimed mostly at db developers (= (99% of time) no one beside Yii db core developers/contributors). It's not an all-purpose solution. It helps to get a required information from the SQL code without getting trapped by regex limitations.
- MySQL's constraint code works with 5.1+. If this is still an issue (tests only pass on 5.0+), I will make it BC with 4.x+/5.0+ (with a help of `SqlTokenizer`).
- `Schema` caching was updated, now it's trivial to add any new information along with the existing without having to think about caching. Tests were intentionally left untouched to catch regressions (if any).

### CODE REVIEW

**Comment by SilverFire** [comment]
Wow, amazing!

**Comment by sergeymakinen** [comment]
I have added `Command` tests and rebased the commits.

**Comment by samdark** [comment]
@sergeymakinen is it ready for review?

**Comment by sergeymakinen** [comment]
@samdark yes, it is!

**Comment by samdark** [comment]
@sergeymakinen have you considered https://github.com/yiisoft/yii2/issues/13647#issuecomment-309023565?

**Comment by sergeymakinen** [comment]
@samdark I did. My current plan is to leave all high level logic in a separate class(-ess). I have started writing tests for this. Pros: it works with any constraint aware Schema/Command classes, it's db independent:

```php
$this->table('test')
    ->column('fk_id')->ifExists()->drop()
    ->alter();
```

Funny thing, it's all still simple. As soon as all tests pass, I'll prepare a PR.

**Comment by sergeymakinen** [comment]
Adding a `dropConstraints()`/`withoutConstraints()->save()` in this case is really trivial.

File: framework/db/Command.php
Comment (samdark) [inline framework/db/Command.php]: Need to explicitly mention in the docs that MySQL simply ignores CHECK constraints.

File: framework/db/ConstraintFinderTrait.php
Comment (samdark) [inline framework/db/ConstraintFinderTrait.php]: metho**d**s

File: framework/db/mysql/Schema.php
Comment (samdark) [inline framework/db/mysql/Schema.php]: Would it work in ANSI mode where quotes are `"`?

**Comment by samdark** [comment]
Overall I love the direction of these changes but I'm a bit concerned about backwards compatibility. What would be broken for the end users at this point?

File: framework/db/mysql/Schema.php
Comment (sergeymakinen) [inline framework/db/mysql/Schema.php]: It should not but [it doesn't seem](https://github.com/sergeymakinen/yii2/blob/3de6f8e397d336d59a7fd8c501a4132631c29563/framework/db/mysql/Schema.php#L200-L214) to bother anyone yet.
Should I fix all of this if it's possible?

File: framework/db/Command.php
Comment (sergeymakinen) [inline framework/db/Command.php]: Sure. I suppose the migration docs should mention all of this new possibilities along with an upcoming table auto column/constraint management. That will be a next part of my work.

File: framework/db/Command.php
Comment (sergeymakinen) [inline framework/db/Command.php]: This migration enchantments cover almost all usable constraints.

I think it _may_ be worth to keep uniques as well as it can be supported (I mean a distinction between unique constraints and unique indexes) in next versions or may be handy when extending core classes.

But I'm not sure. I still can't way any (useful in PHP) db which doesn't create an index behind the constraint. (NoSQL?..)

I would like to hear opinions if it should go or stay.

**Comment by sergeymakinen** [comment]
@samdark thanks for the review! It should not affect a current code. I redesigned a table metadata caching but kept old tests and even added a version in a cache bucket to exclude caching issues. Other things are additions and they don't touch any existing code.

Existing methods to retrieve constraint/column info should be replaced/updated to use this new code but, well, 6k lines changed, so I suggest to review and merge this code, then I'll roll out a PR with a new table builder, then I'll start updating docs and prepare another PR to replace/update old methods but only where BC is safe and secure.

File: framework/db/Command.php
Comment (samdark) [inline framework/db/Command.php]: Isn't PostgreSQL supporting both?

File: framework/db/Command.php
Comment (samdark) [inline framework/db/Command.php]: This one isn't about migration docs. It's about phpdoc.

File: framework/db/mysql/Schema.php
Comment (samdark) [inline framework/db/mysql/Schema.php]: Yes, I think so.

Comment (samdark) [inline framework/db/mysql/Schema.php]: Maybe in a separate PR. If current PR doesn't support ANSI quotes and noone complains then probably it's fine to have it as is.

File: framework/db/Command.php
Comment (sergeymakinen) [inline framework/db/Command.php]: > PostgreSQL automatically creates a unique index when a unique constraint or primary key is defined for a table.

It works the [same](https://www.postgresql.org/docs/9.4/static/indexes-unique.html). Em.

File: framework/db/Command.php
Comment (sergeymakinen) [inline framework/db/Command.php]: Oh. Okay. :)

File: framework/db/mysql/Schema.php
Comment (sergeymakinen) [inline framework/db/mysql/Schema.php]: Then I'll do it later to avoid unneeded rebasing and merging.

File: framework/db/Command.php
Comment (samdark) [inline framework/db/Command.php]: OK. Let's have it.

File: framework/db/ConstraintFinderTrait.php
Comment (dynasource) [inline framework/db/ConstraintFinderTrait.php]: shouldn't we use `constants` for strings linke 'primaryKey'?

File: framework/db/Schema.php
Comment (dynasource) [inline framework/db/Schema.php]: mixed? Additionally, isnt key a bit confusing as a name for returning an array?

**Review by dynasource — CHANGES_REQUESTED** [review]
impressive effort. For what I've checked, it looks good. With respect to BC, I tested the branch in one my projects and all tests pass in Sqlite and Mysql, so thats a good signal.

File: framework/db/Command.php
Comment (samdark) [inline framework/db/Command.php]: I think @dynasource suggestion makes perfect sense.

File: framework/db/ConstraintFinderTrait.php
Comment (samdark) [inline framework/db/ConstraintFinderTrait.php]: Constants could not be private till PHP 7.2. Exposing implementation details isn't a good idea IMO.

File: framework/db/Command.php
Comment (sergeymakinen) [inline framework/db/Command.php]: It sounds reasonable, but here are thoughts behind `addUnique`:
1. Almost all `Command` / `QueryBuilder` names follow SQL commands.
2. We `CREATE INDEX`es but we `ADD CONSTRAINT`s when we `ALTER TABLE`.
3. `Unique` isn't ideal, but why don't we use `Constraint` suffix elsewhere then (hint: BC)? I prefer to stick with consistency - like @dynasource said.

But I'm not principled in this case.

File: framework/db/Schema.php
Comment (sergeymakinen) [inline framework/db/Schema.php]: It's because a key can be [anything](https://github.com/yiisoft/yii2/blob/master/framework/caching/Cache.php#L93) JSON serializable.
I can't change it because it's `protected`, the return type is wide (`mixed`) and we're in 2.0.*

File: framework/db/Constraint.php
Comment (samdark) [inline framework/db/Constraint.php]: Wouldn't it make sense to default to `[]` instead of `null` here?

File: framework/db/IndexConstraint.php
Comment (samdark) [inline framework/db/IndexConstraint.php]: Shouldn't it be `false` by default?

File: framework/db/IndexConstraint.php
Comment (samdark) [inline framework/db/IndexConstraint.php]: Shouldn't it be `false` by default?

File: framework/db/Schema.php
Comment (samdark) [inline framework/db/Schema.php]: Need phpdoc.

File: framework/db/CheckConstraint.php
Comment (samdark) [inline framework/db/CheckConstraint.php]: Alternatively, MySQL schema may just throw NotSupportedException since, in fact, CHECK is not supported while syntax is accepted.

File: framework/db/Constraint.php
Comment (sergeymakinen) [inline framework/db/Constraint.php]: `null` here and in other places means "we got no info".

File: framework/db/CheckConstraint.php
Comment (sergeymakinen) [inline framework/db/CheckConstraint.php]: I do agree with `NotSupportedException`. I will check other cases as well.

File: framework/db/Constraint.php
Comment (samdark) [inline framework/db/Constraint.php]: OK.

**Comment by sergeymakinen** [comment]
Probably most issues are addressed.

File: framework/db/SqlToken.php
Comment (samdark) [inline framework/db/SqlToken.php]: Looks suspicious. You're incrementing children index but accessing parent with it. Is it intended?

File: framework/db/SqlToken.php
Comment (samdark) [inline framework/db/SqlToken.php]: Same here. You're incrementing children index but accessing parent with it. Is it intended?

File: framework/db/mssql/Schema.php
Comment (samdark) [inline framework/db/mssql/Schema.php]: Else isn't needed.

File: framework/db/mysql/Schema.php
Comment (samdark) [inline framework/db/mysql/Schema.php]: Else isn't needed.

File: framework/db/oci/Schema.php
Comment (samdark) [inline framework/db/oci/Schema.php]: Else isn't needed.

**Review by samdark — CHANGES_REQUESTED** [review]

File: framework/db/Schema.php
Comment (sergeymakinen) [inline framework/db/Schema.php]: Will it affect BC?

Comment (samdark) [inline framework/db/Schema.php]: It could. Let's leave it as is for now.

File: framework/db/SqlToken.php
Comment (sergeymakinen) [inline framework/db/SqlToken.php]: It is: `ArrayAccess` provides access to the `children` property with a handy ability to specify negative offsets which makes the code shorter and easier to understand.
Basically we iterate token by token with an exception of `any` that toggles an iteration until we matched with a next pattern token or EOF.
Is it clear or should I add this description before `foreach`? As 3 months passed by I found this description useful 'cause it took some time to remember…

File: framework/db/ConstraintFinderTrait.php
Comment (sergeymakinen) [inline framework/db/ConstraintFinderTrait.php]: > Note:
As of PHP 7.1.0 visibility modifiers are allowed for class constants.

[Source](http://php.net/manual/en/language.oop5.constants.php#language.oop5.basic.class.this)

I think it's a good idea to constantify those magic strings in 2.1.

File: framework/db/SqlToken.php
Comment (samdark) [inline framework/db/SqlToken.php]: Yes, a comment would be helpful in this case, I think. Wasn't easy to get it.

File: framework/db/ConstraintFinderTrait.php
Comment (samdark) [inline framework/db/ConstraintFinderTrait.php]: Yes.

File: framework/db/SqlToken.php
Comment (sergeymakinen) [inline framework/db/SqlToken.php]: AFAIK there are no phpdoc requirements for `private`s, em?

Comment (samdark) [inline framework/db/SqlToken.php]: These should be described as well.

Comment (sergeymakinen) [inline framework/db/SqlToken.php]: I don't think there's any need for param docs as I covered it in the public method. But anyway take a look at it again after a commit is pushed.

**Comment by sergeymakinen** [comment]
@samdark okay, your turn! :)

**Review by samdark — APPROVED** [review]

**Comment by sergeymakinen** [comment]
Just a side note: while working my high-level extension I figured out there's no clean way to check if a schema supports constraints if we don't introduce an interface for them. Extracting public methods from `ConstraintFinderTrait` to `ConstraintAwareInterface` would be a good idea IMO.

Similar case with views but I believe they should be a part of the core `Schema` in 2.1 (and it's another issue).

**Comment by samdark** [comment]
Do you think this pull request should include this interface?

**Comment by sergeymakinen** [comment]
IMO we can include in a next PR which will introduce this table migrator along with some new DMBS properties in `Schema` (like `$maxIdentifierLength`).

**Comment by samdark** [comment]
OK. I'm fine merging it. Waiting for approval from @klimov-paul, @SilverFire, @cebe, @dynasource.

File: framework/db/cubrid/Schema.php
Comment (cebe) [inline framework/db/cubrid/Schema.php]: variable `$columns` is not defined here? please make sure this is covered by a test. A simple one is sufficient for now.

**Review by cebe — CHANGES_REQUESTED** [review]
Looks good to me overall, thanks for the huge effort!

I have some minor remarks, see comments I made. Also, a CHANGELOG entry is missing :)

Also two questions:

- why is default value a constraint?
- why is index a constraint?

**Comment by cebe** [comment]
@sergeymakinen made some minor changes and pushed to your branch.

File: framework/db/SqlToken.php
Comment (sergeymakinen) [inline framework/db/SqlToken.php]: SQLite's one is only implemented ATM so it's generic for now. We can change it later.

File: framework/db/ConstraintFinderTrait.php
Comment (sergeymakinen) [inline framework/db/ConstraintFinderTrait.php]: It's called (automagically) by `getTableMetadata` & `getSchemaMetadata`. I forgot to mention it in the method descriptions. Will fix it.

File: framework/db/ConstraintFinderTrait.php
Comment (sergeymakinen) [inline framework/db/ConstraintFinderTrait.php]: Do you mean by replacing abstracts with `NotSupportedException` (to maintain BC)? It's possible. The only reason behind this trait was to possibly make `Schema` look "lighter" if someone doesn't need a constraint support at all.
Maybe it's just an overcomplication and we simply need to put them all in `Schema`. I even suggested the same for `ViewFinderTrait`… I don't know TBH. :)
cc @samdark

File: framework/db/ConstraintFinderTrait.php
Comment (sergeymakinen) [inline framework/db/ConstraintFinderTrait.php]: Sounds reasonable but let's decide on merging `ConstraintFinderTrait` with `Schema`. Maybe these problems will be auto solved.

File: framework/db/cubrid/Schema.php
Comment (sergeymakinen) [inline framework/db/cubrid/Schema.php]: It actually is covered… I just wanted to be sure an array data is really sorted if something changes in next versions and CUBRID will start returning array values in random order. Thanks for spotting! :)

**Comment by sergeymakinen** [comment]
@cebe thank you for the review! I fixed/answered the questions. We need to decide if we really need a trait or not. If not, if makes `Schema` larger but easier to work with. And it will eliminate the need for the additional interface at all.

**Comment by sergeymakinen** [comment]
Oh, how dedicated questions:
1. Because defaults are constraints in some DBMS. Like MSSQL.
2. Because uniques constraints are usually "powered up" by indexes or vice versa. So they're tied together and it's easier to call an index a constraint and work with them the same way.

File: framework/db/SqlToken.php
Comment (cebe) [inline framework/db/SqlToken.php]: Changing it later might need to break BC so I'd prefer to change it now. how about passing a SqlToken directly so the tokenizing is made outside of the matches() call?

File: framework/db/ConstraintFinderTrait.php
Comment (cebe) [inline framework/db/ConstraintFinderTrait.php]: its fine to have them abstract in Schema, but I see the reasoning about allowing to implement Schema without the need to implement the trait. It makes sense and avoids BC break with custom DB implementations. Looks good to me. :shipit:

File: framework/db/SqlToken.php
Comment (sergeymakinen) [inline framework/db/SqlToken.php]: It sounds good to me. :)

**Comment by cebe** [comment]
@sergeymakinen I proposed some changes in https://github.com/sergeymakinen/yii2/pull/1

**Comment by sergeymakinen** [comment]
@cebe merged, thanks! :)

**Comment by samdark** [comment]
@sergeymakinen please add a line for changelog.

**Review by cebe — APPROVED** [review]

**Comment by cebe** [comment]
Merged, thank you very much!

**Comment by sergeymakinen** [comment]
Thanks everyone for reviews! And see you in a next PR. ;)

**Pull request merged**
